src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2023 Monica Omar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Monica Omar
+-/
+import linear_algebra.basic
+import linear_algebra.projection
+
+/-!
+# Invariant submodules
+
+In this file, we define and prove some basic results on invariant submodules.
+-/
+
+namespace submodule
+
+variables {E R : Type*} [ring R] [add_comm_group E] [module R E]
+
+/-- `U` is `T` invariant : `U ≤ U.comap` -/
+def invariant_under (U : submodule R E) (T : E →ₗ[R] E) : Prop := U ≤ U.comap T
```

Comment:
Would it be beneficial to define this for a `set (E →ₗ[R] E)`? It would then be possible to refactor [representation.invariants](https://leanprover-community.github.io/mathlib_docs/representation_theory/invariants.html#representation.invariants) to use this in the future, for example.

Reply:
If we're going down that route we should probably also define it for plain functions rather than `linear_map`s

Reply:
Wouldn't be a submodule for general functions.

Reply:
Oh sorry, I thought you were suggesting
```lean
/-- `U` is `T` invariant : `U ≤ U.preimage T` -/
def invariant_under (U : set E) (T : E → E) : Prop := U ≤ U.preimage T
```
(with `\subset` and `\preimage` notation)

Reply:
Oh sorry, I had a serious misconception: the invariance in this PR is the submodule being mapped into itself under T, rather than being fixed pointwise by T. Please forget all what I've said!

Reply:
By defining it for a `set (E →ₗ[R] E)`, I thought you meant that a submodule is invariant under each linear map in the set. Because that might come in handy in certain situations - but it's probably better to add a definition for that separately.